### PR TITLE
[expression] Add line_interpolate_point_by_m and line_locate_m expressions

### PR DESCRIPTION
**Pull request by nirvn**
## Description

This PR adds two new functions to our expression engine to work with M values along line strings:

- `line_interpolate_point_by_m`: returns a point geometry of a matching m value interpolated along a line containing an m dimension. 
- `line_locate_m`: returns a distance from the beginning of a line where a matching m value was found.

These two functions coming in extremely handy when working with temporal data, e.g. a linestring representing a GPS track where the M value represents the epoch value.

Using this function alongside of temporal controller makes for beautiful animations that were previously harder to unlock:

https://github.com/user-attachments/assets/8a062173-dbf4-4af4-9173-210cf957820c

### CODE REVIEW

File: src/core/expression/qgsexpressionfunction.cpp [hunk 1]
```diff
@@ -5653,6 +5653,33 @@ static QVariant fcnLineInterpolatePoint( const QVariantList &values, const QgsEx
...
+  QgsGeometry lineGeom = QgsExpressionUtils::getGeometry( values.at( 0 ), parent );
+  const double m = QgsExpressionUtils::getDoubleValue( values.at( 1 ), parent );
+  const bool use3DDistance = values.at( 2 ).toBool();
+
+  double x, y, z, distance;
+
+  const QgsLineString *line = qgsgeometry_cast<const QgsLineString *>( lineGeom.constGet() );
+  if ( !line )
+  {
+    return QVariant();
```
Comment (troopa81) [inline src/core/expression/qgsexpressionfunction.cpp:5667]: Shakll we not set an evaluation error string?

File: src/core/expression/qgsexpressionfunction.cpp [hunk 2]
```diff
@@ -5653,6 +5653,33 @@ static QVariant fcnLineInterpolatePoint( const QVariantList &values, const QgsEx
   return result;
 }
 
+static QVariant fcnLineInterpolatePointByM( const QVariantList &values, const QgsExpressionContext *, QgsExpression *parent, const QgsExpressionNodeFunction * )
+{
+  QgsGeometry lineGeom = QgsExpressionUtils::getGeometry( values.at( 0 ), parent );
```
Comment (troopa81) [inline src/core/expression/qgsexpressionfunction.cpp:5658]:
```suggestion
  const QgsGeometry lineGeom = QgsExpressionUtils::getGeometry( values.at( 0 ), parent );
```

File: src/core/expression/qgsexpressionfunction.cpp [hunk 3]
```diff
@@ -5732,6 +5759,24 @@ static QVariant fcnLineLocatePoint( const QVariantList &values, const QgsExpress
   return distance >= 0 ? distance : QVariant();
 }
 
+static QVariant fcnLineLocateM( const QVariantList &values, const QgsExpressionContext *, QgsExpression *parent, const QgsExpressionNodeFunction * )
+{
+  QgsGeometry lineGeom = QgsExpressionUtils::getGeometry( values.at( 0 ), parent );
```
Comment (troopa81) [inline src/core/expression/qgsexpressionfunction.cpp:5764]:
```suggestion
  const QgsGeometry lineGeom = QgsExpressionUtils::getGeometry( values.at( 0 ), parent );
```

File: src/core/expression/qgsexpressionfunction.cpp [hunk 4]
```diff
@@ -5732,6 +5759,24 @@ static QVariant fcnLineLocatePoint( const QVariantList &values, const QgsExpress
...
+  const double m = QgsExpressionUtils::getDoubleValue( values.at( 1 ), parent );
+  const bool use3DDistance = values.at( 2 ).toBool();
+
+  double x, y, z, distance;
+
+  const QgsLineString *line = qgsgeometry_cast<const QgsLineString *>( lineGeom.constGet() );
+  if ( !line )
+  {
+    return QVariant();
```
Comment (troopa81) [inline src/core/expression/qgsexpressionfunction.cpp:5773]: error eval message?

File: resources/function_help/json/line_interpolate_point_by_m [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  "description": "Returns the point interpolated by a matching M value along a linestring geometry.",
+  "arguments": [{
+    "arg": "geometry",
+    "description": "a linestring geometry"
+  }, {
+    "arg": "m",
+    "description": "an M value"
+  }, {
+    "arg": "use_3d_distance",
+    "optional": true,
```
[the remark this answers is not in the document]
Comment (nirvn) [reply resources/function_help/json/line_interpolate_point_by_m:14]: Done

File: src/core/expression/qgsexpressionfunction.cpp (hunk 2 above)
Comment (nirvn) [reply src/core/expression/qgsexpressionfunction.cpp:5658]: Done

File: src/core/expression/qgsexpressionfunction.cpp (hunk 1 above)
Comment (nirvn) [reply src/core/expression/qgsexpressionfunction.cpp:5667]: I think we should match the behavior to the pre-existing line_interpolate_point(), which does not return an evaluation error string.

File: src/core/expression/qgsexpressionfunction.cpp (hunk 3 above)
Comment (nirvn) [reply src/core/expression/qgsexpressionfunction.cpp:5764]: Done.

File: src/core/expression/qgsexpressionfunction.cpp (hunk 4 above)
Comment (nirvn) [reply src/core/expression/qgsexpressionfunction.cpp:5773]: I think we should match the behavior to the pre-existing line_locate_point() which does not return an evaluation error string.

**Comment by nirvn** [comment]
@DelazJ , @troopa81 , thanks for the review, comments addressed / justified :)

**Review by nyalldawson — APPROVED** [review]

**Comment by nirvn** [comment]
@troopa81 , merci.

**Pull request merged**